text/0000-ember-data-singleton-record-data.md

```diff
@@ -0,0 +1,450 @@
+
+# Singleton Record Data
+
+## Summary
+
+Ensures that `RecordData` can be implemented as a singleton, eliminates several redundant APIs
+that creeped into the original implementation, and simplifies the method signatures of
+RecordData APIs by using [`Identifiers`](https://github.com/emberjs/rfcs/pull/403).
+
+## Motivation
+
+`RecordData` is the data-cache primitive for information given to `ember-data`, and while
+a cache-per-entity setup is sometimes desired, a singleton cache offers opportunities for
+performance optimization and improved feature sets. Current default `RecordData`, and
+`InternalModel` which it replaced are examples of cache-per-entity strategies.
+
+Our original intent [when discussing `RecordData`](https://github.com/emberjs/rfcs/pull/293)
+was for it to be possible to be implemented as a singleton; however, this intent was not
+captured well in that RFC and while the APIs presented there would have enabled it, the
+actual implementation differed in ways that prevent `singleton` implementations.
+
+The introduction of `Identifiers` presents us with a good opportunity to refactor the API
+surface of our cache-primitive, simplifying and streamlining how it works while solidifying
+and codifying its ability to be a `singleton`.
+
+### Reduce Overloading
+
+Previously, to deal with the lack of a unified identifier concept, we overloaded many
+`StoreWrapper` and `RecordData` method signatures with `modelName, id, clientId` as
+`arguments`. The original intent was to overload **all** of their method signatures
+in this way to ensure `singleton` implementations could be built, but we failed to
+correctly implement the original `RecordData RFC` in this regard.
+
+The introduction of `Identifiers` provides us a cleaner interface for communicating identity
+in these APIs. For uniformity, methods will always take an `identifier` as their **first
+argument**.
+
+## Detailed design
+
+Because `RecordData` is a public userland interface we
+must rely on capabilities reporting to handle the
+deprecations for it introduced by this RFC.
+
+We will use this opportunity to improve the encapsulation
+of `RecordData` via introduction of a `RecordDataManager`.
+
+Each new `RecordData` instance will be paired with an
+instance of a `RecordDataManager`. The manager instance will
+be responsible for handling interop, deprecations and other
+book-keeping tasks we require as needed.
+
+RecordData implementations must now specify a `version`.
+The implicit version when unspecified is `"1"` (the version
+which this RFC supplants and deprecates). Implementations
+without a version or with a version equal to `"1"` will
+receive a deprecation notice the first time an instance of
+a previously unseen `RecordData` class is encountered.
+
+Note that to comply with `SemVer`, any given release of `ember-data` will
+only have support for a maximum of two versions of `RecordData`, with
+deprecated versions being dropped following LTS releases.
```

Comment:
This is a good policy and standard in the ember ecosystem, but it doesn't really have much to do with semver per se.

Also, it's somewhat implied, but we should be explicit that we only intend to increment these versions at LTS, so that "we support 2 versions of RecordData" means in practice that we support the versions that existed at LTS versions of ember data.

Reply:
Maybe something like


> In keeping with the support policy of Ember.js, future releases of `ember-data` will support the versions of RecordData current at the time of the last two `ember-data` LTS releases

Reply:
I like that wording